cmd/snap-confine/snap-confine-args.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "snap-confine-args.h"
+
+#include <string.h>
+
+#include "../libsnap-confine-private/utils.h"
+
+struct sc_args {
+	// The security tag that the application is intended to run with
+	char *security_tag;
+	// The executable that should be invoked
+	char *executable;
+
+	// Flag indicating that --version was passed on command line.
+	bool is_version_query;
+	// Flag indicating that --classic was passed on command line.
+	bool is_classic_confinement;
+};
+
+struct sc_args *sc_nonfatal_parse_args(int *argcp, char ***argvp,
+				       struct sc_error **errorp)
+{
+	struct sc_args *args = NULL;
+	struct sc_error *err = NULL;
+
+	if (argcp == NULL || argvp == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argcp or argvp is NULL");
+		goto out;
+	}
+	// Use dereferenced versions of argcp and argvp for convenience.
+	int argc = *argcp;
+	char **const argv = *argvp;
+
+	if (argc == 0 || argv == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc is zero or argv is NULL");
+		goto out;
+	}
+	// Sanity check, look for NULL argv entries.
```

Comment:
Interesting observation. This was in response to a question from @chipaca (that we don't check for NULL arguments). I think you are right. Since we don't (and cannot really) check argc in a reasonable way, checking argv is equally unreasonable.

I was looking if there's a way to attack snap-confine by exec'ing it with a malicious argument vector. Suggestions are very much welcome.